### PR TITLE
Perf/batched trie visitor

### PR DESCRIPTION
**Pull request by asdacap**
- Create a different trie node visitor that attempt to reduce iops requirement by reordering reads according to key. The gains depends on how much key is in the memory which is configurable.
- In theory, the raw size of the working set for this to be reasonably effective is about 256MB for goerli and 1GB for mainnet.
- On goerli with 2GB memory budget, this increase the rate of copy by a factor of two. Similar result with mainnet but with 4GB memory budget. On mainnet, if you give it as much RAM as it want, it take close to 4 times the standard rate and complete in 45 minute. RSS is 30GB, GC stats at 10GB.  Note: I have 128GB of ram and the whole goerli db can fit in it.
- A more reasonable mainnet configuration is 4GB memory budget which complete the pruning in about 1 hour 10 minute. RSS is 18.5GB with GC stats at 3GB.

(before, after)
Screenshot from 2023-03-22 04-05-13

## Changes

- _List the changes_

#### What types of changes does your code introduce?

- [X] Optimization

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Notes on testing

- I had a goerli db on its 22th state db.

### CODE REVIEW

**Comment by kamilchodola** [comment]
@asdacap 
Can You please throw some commands to be applied before start of pruning? Or is it default settings thing?

File: src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+        // This need to be very small
+        _partitions = new CompactStack<Job>[_partitionCount];
+        for (int i = 0; i < _partitionCount; i++)
+        {
+            _partitions[i] = new CompactStack<Job>(jobPool);
+        }
+    }
+
+    // Determine the locality of the key. I guess if you use paprika or something, you'd need to modify this.
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs:101]: @Scooletz :)

File: src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+            if (_activeJobs == 0 || _failed)
+            {
+                // Its finished
+                return null;
+            }
+
+            if (_queuedJobs == 0)
+            {
+                // Just a small timeout to prevent threads from loading CPU
+                // Note, there are other threads also going through the stacks, so its fine to have this high.
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs:156]: btw have you checked how this full pruning method affects block processing? That would be something interesting from validator's perspective
If not, we should check it during next experiments (cc: @kamilchodola )

File: src/Nethermind/Nethermind.Runner/configs/mainnet.cfg [hunk 3]
```diff
@@ -35,7 +35,10 @@
       "http://localhost:8551|http;ws|net;eth;subscribe;engine;web3;client"
     ]
   },
+  "Pruning": {
+    "FullPruningMemoryBudgetMb": 4000
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Runner/configs/mainnet.cfg:39]: From my grafana, the RSS (memory usage of whole process) is about 19GB. The stats from GC (gen0+1+2+LOH) is about 2 to 3GB. The actuall target is half the memory budget, so the 2 to 3GB GC stats is expected.

File: src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs:101]: I don't think (and I could be wrong here), it would benefit paprika. As the nodes that is written at around the same time wil naturally have similar offset with paprika.

File: src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs:156]: It quite a lot worst. Likely due to higher rate of processing. Seems to load CPU quite heavily.

File: src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        // The keccak + context itself should be 40 byte. But the measured byte seems to be 52 from GC stats POV.
+        // The * 2 is just margin. RSS is still higher though, but that could be due to more deserialization.
+        long recordSize = 52 * 2;
+        long recordCount = visitingOptions.FullScanMemoryBudget / recordSize;
+
+        // Generally, at first, we want to attempt to maximize number of partition. This tend to increase throughput
+        // compared to increasing batch size.
+        _partitionCount = recordCount / _maxBatchSize;
+
+        // 3000 is about the num of file for state on mainnet, so we assume 4000 for an unpruned db. Multiplied by
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs:71]: so if I understand the comment you chose the best value for mainnet - 4000. However, what about other chains? Should we make it configurable per chain? Of course, mainnet >>>> all ethereum testnets. However, good support for Gnosis chain is expected.

**Review by MarekM25 — COMMENTED** [review]
Amazing work. Added a few comments, mostly questions.

Tl;dr I think we could merge it now, but with `FullScanMemoryBudget=0`

My main concerns are:
-users with 16GB ram that are used to the previous method of pruning (not ram consuming)
-block processing performance (validator performance) during the full pruning, question mark, but I guess it might be much worse than on normal full pruning, but of course, it will finish faster

File: src/Nethermind/Nethermind.State/IStateProviderExtensions.cs [hunk 5]
```diff
@@ -26,7 +27,11 @@ public static string DumpState(this IStateProvider stateProvider)
         public static TrieStats CollectStats(this IStateProvider stateProvider, IKeyValueStore codeStorage, ILogManager logManager)
         {
             TrieStatsCollector collector = new(codeStorage, logManager);
-            stateProvider.Accept(collector, stateProvider.StateRoot, new VisitingOptions { MaxDegreeOfParallelism = Environment.ProcessorCount, });
+            stateProvider.Accept(collector, stateProvider.StateRoot, new VisitingOptions
+            {
+                MaxDegreeOfParallelism = Environment.ProcessorCount,
+                FullScanMemoryBudget = 16.GiB(), // Gonna guess that if you are running this, you have a decent setup.
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.State/IStateProviderExtensions.cs:33]: Nah, If you are using this, you are a dev and you'd probably can change this in code manually.

File: src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+            {
+                // Its finished
+                return null;
+            }
+
+            if (_queuedJobs == 0)
+            {
+                // Just a small timeout to prevent threads from loading CPU
+                // Note, there are other threads also going through the stacks, so its fine to have this high.
+                Thread.Sleep(TimeSpan.FromMilliseconds(100));
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs:157]: I'm not even sure if this is working. This is only useful way at the end of the prune. It'll just load the CPU. I'd recommend just setting the degree of parallelism lower if user want to prioritize block processing. The default all core is likely not useful.

**Comment by kamilchodola** [comment]
Started a node with name:
NodeName='DevNode-kamilchodola-perf-batched-trie-visitor-mainnet-prysm'
With params:
Pruning.FullPruningMinimumDelayHours=0 Pruning.FullPruningThresholdMb=100000 Pruning.FullPruningTrigger=StateDB Pruning.Mode=Full

For 48h.
Will let You know how it looks and how many times pruning executed during that time.

File: src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs:157]: Most of our users (90% or more) would like to prioritize block processing over full pruning. I'm curious if we should change the default full pruning values then. Obviously, out of the scope of the PR

File: src/Nethermind/Nethermind.Trie/TrieNodeResolverWithReadFlags.cs [hunk 7]
```diff
@@ -23,8 +23,13 @@ public TrieNode FindCachedOrUnknown(Keccak hash)
         return _baseResolver.FindCachedOrUnknown(hash);
     }
 
-    public byte[]? LoadRlp(Keccak hash)
+    public byte[]? LoadRlp(Keccak hash, ReadFlags flags = ReadFlags.None)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Trie/TrieNodeResolverWithReadFlags.cs:26]: Cosmetic - change name from flags to overridenFlags? Wdyt?

File: src/Nethermind/Nethermind.Trie/PatriciaTree.cs [hunk 8]
```diff
@@ -1030,14 +1030,14 @@ public void Accept(ITreeVisitor visitor, Keccak rootHash, VisitingOptions? visit
                 resolver = new TrieNodeResolverWithReadFlags(TrieStore, ReadFlags.HintCacheMiss);
             }
 
+            visitor.VisitTree(rootHash, trieVisitContext);
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/PatriciaTree.cs:1033]: batchedTrieVisitor does not call this at all, so this need to be moved so that the timer works correctly.

File: src/Nethermind/Nethermind.Trie/TrieNodeResolverWithReadFlags.cs (hunk 7 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Trie/TrieNodeResolverWithReadFlags.cs:26]: Can change `_flags` to `_defaultFlags` how about that?

Comment (MarekM25) [reply src/Nethermind/Nethermind.Trie/TrieNodeResolverWithReadFlags.cs:26]: Much better. I've already changed during rereview :)

**Review by MarekM25 — APPROVED** [review]

File: src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs [hunk 9]
```diff
@@ -171,7 +172,11 @@ protected virtual void RunPruning(IPruningContext pruning, Keccak statRoot)
             {
                 pruning.MarkStart();
                 using CopyTreeVisitor copyTreeVisitor = new(pruning, _logManager);
-                VisitingOptions visitingOptions = new() { MaxDegreeOfParallelism = _pruningConfig.FullPruningMaxDegreeOfParallelism };
+                VisitingOptions visitingOptions = new()
+                {
+                    MaxDegreeOfParallelism = _pruningConfig.FullPruningMaxDegreeOfParallelism,
+                    FullScanMemoryBudget = ((long)_pruningConfig.FullPruningMemoryBudgetMb).MiB(),
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Blockchain/FullPruning/FullPruner.cs:178]: This going to be x 1024*1024 too high? If it comes in as MB already then is multiplied by `MiB`

File: src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+            for (int i = 0; i < resolveOrdering.Count; i++)
+            {
+                int idx = resolveOrdering[i];
+
+                (TrieNode nodeToResolve, SmallTrieVisitContext ctx) = currentBatch[idx];
+                try
+                {
+                    Keccak theKeccak = nodeToResolve.Keccak;
+                    nodeToResolve.ResolveNode(_resolver);
+                    nodeToResolve.Keccak = theKeccak; // The resolve may set a key which clear the keccak
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs:318]: Keccak disappearing could cause issues for logic that saw it was there; can a flag be passed to make sure it doesn't get cleared by this?
